### DISCUSSION TITLE
Custom Service name provider for Java

### DISCUSSION THREAD

**Discussion by mortada87** [open-telemetry/opentelemetry-java-instrumentation#7966] [Q&A]
Hello, 
I need to be able to dynamically set the service name  for my Opentelemetry java agent, so that it takes the context path of the webapp, knowing that , multiple wars run on the same tomcat instance.
thanks for your help and for the great job done on this subject.
Hamid

**Comment by trask** [comment]
hi @mortada87 this may be what you are looking for: https://github.com/open-telemetry/opentelemetry-java-contrib/pull/562

cc @breedx-splk

**Reply by mortada87** [reply]
Hi @trask , thanks for your answer, I think that's what I am looking for , I wonder however how can I instruct the javagent jar how to use this lib?
Thanks a lot, huge fan of Glowroot btw ;)
Hamid

**Reply by mateuszrzeszutek** [reply]
You can simply load the jar with that lib as a [javaagent extension](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension)

**Reply by breedx-splk** [reply]
> hi @mortada87 this may be what you are looking for: [open-telemetry/opentelemetry-java-contrib#562](https://github.com/open-telemetry/opentelemetry-java-contrib/pull/562)

I don't think that's what's being asked. I think the user is asking for a dynamic service name, so that server requests coming in on `/foo/bar` have `service.name=fooservice` but server requests coming in on the same tomcat instance to a different war on `/bar/baz` have `service.name=barservice`. At least that's the way I read the request.

As far as I know, the OpenTelemetry agent doesn't do this. A little context -- the `service.name` is part of the "[service](https://github.com/open-telemetry/opentelemetry-specification/blob/431d7d5e413a334d89396161f48cc7d82f445ba7/specification/resource/semantic_conventions/README.md#service)" which is part of the [resource](https://github.com/open-telemetry/opentelemetry-specification/blob/431d7d5e413a334d89396161f48cc7d82f445ba7/specification/resource/sdk.md). The resource is definitely immutable, and the instrumentation applies to the entire JVM, not individual applications. 

On the implementation side of the SDK, the immutable resource is passed to the `TracerProvider` and SDK only has a single `TracerProvider`. As such, this kinda means only ever having one `Resource` per SDK instance.

I also suspect that this mode of deploying multiple webapps to a single JVM has fallen out of fashion, largely due to containerization.

**Reply by mortada87** [reply]
Hello @breedx-splk ,
That's exactly what Im looking for. I guess that's not possible, i cannot write something myself to do that?
Thanks,
Hamid

**Reply by mortada87** [reply]
> You can simply load the jar with that lib as a [javaagent extension](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension)

Thanks a lot

**Reply by trask** [reply]
> i cannot write something myself to do that?

it's not straight forward, but it is possible, at a high level:

* write a SpanExporter wrapper, and "mutate" span resources based on their `http.target`
* in order to make this work for child spans also, you would need to pass the `http.target` down from the parent spans to the child spans so that it is available in your SpanExporter wrapper

**Reply by mortada87** [reply]
Hello Trask, thanks a lot, it would be something like this?

`
public void export(Collection<SpanData> spanDataCollection)
{
spanDataCollection.stream().map(spanData -> spanData.getResource().merge(Resource.create(Attributes.of(ResourceAttributes.SERVICE_NAME,TracerHelper.getApplicationame()))));
}
`

Thanks,

Hamid

**Reply by trask** [reply]
SpanData is immutable, so you'll probably need to put the merged resource into a "new" SpanData, e.g. using https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/data/DelegatingSpanData.java

**Reply by mortada87** [reply]
Thanks a lot for your help! Will give it a try.
